src/nc_core.c

```diff
@@ -272,14 +338,14 @@ core_timeout(struct context *ctx)
         msg = msg_tmo_min();
         if (msg == NULL) {
             ctx->timeout = ctx->max_timeout;
-            return;
+            break;
         }
 
         /* skip over req that are in-error or done */
 
         if (msg->error || msg->done) {
             msg_tmo_delete(msg);
-            continue;
+            break;
```

Comment:
What's the intent of changing this from a `continue` to a `break` - It seems to work anyway, but if there are a large number of timeouts should this try to process those first for efficiency and close connections by continuing?

Should the return below also be a break;?

Reply:
This also seems possibly concerning - what would happen if these still belonged to a connection (e.g. multiple heartbeats), and we retried a connection before processing some requests that had also errored on that connection.

I guess that `break;` may make it less likely to call retry_connection on the same server twice, though that should be a different check.

I see that retry_connection will call `add_failed_server`. Should `add_failed_server` iterate over the array to check if something already marked the server as failed (e.g. if multiple heartbeats were sent and got multiple errors, e.g. a server that was connected to but immediately resulted in a protocol error)